web/src/bootstrap_typeahead.ts

```diff
@@ -375,7 +377,24 @@ export class Typeahead<ItemType extends string | object> {
             // We expect the typeahead creator to handle when to hide / show the typeahead.
             trigger: "manual",
             arrow: false,
-            offset: [0, 0],
+            offset({placement}) {
+                if (input_element.type === "textarea") {
+                    const caret = getCaretCoordinates(
+                        input_element.$element[0],
+                        input_element.$element[0].selectionStart,
+                    );
+                    if (placement === "top-start") {
+                        return [caret.left, -caret.top + 4];
+                    }
+                    if (placement === "bottom-start") {
+                        const field_height = input_element.$element.height();
+                        assert(field_height, "Text area height property missing");
+                        const distance = field_height - caret.top - 8;
+                        return [caret.left, -distance];
+                    }
```

Comment:
Should we just move the typeahead that uses `bottom-start` to use `top-start` too and remove this logic?

Reply:
The caret co-ordinates are relative to top left corner of text area. In `bottom-start` the offset works from bottom left corner. Thus the height of reference (text-area) is used to calculate the offset from top left corner.

Reply:
If we remove the `bottom-start` and only keep `top-start`, there's an overflow case which will arise due to removing the fallback case.

![image](https://github.com/zulip/zulip/assets/78212328/abc05fe5-1c2a-45c0-bc70-277eb5ea5084)

Reply:
Oh yeah! Makes sense to have this case anyway then. I think my only desired tweak would be use a variable for 4 here to better represent what this stands for. Otherwise, LGTM.

Reply:
The 4 is to add a slight gap between typeahead box and caret.

This is with gap:

![image](https://github.com/zulip/zulip/assets/78212328/e295e0d9-4bf1-48d8-bf0d-98a9f8c25ad9)

This is without gap: (feels overlapped)

![image](https://github.com/zulip/zulip/assets/78212328/15e570d6-8edd-45ec-aafb-fe4ff79ec79d)

Reply:
Added a variable for it

Reply:
Maybe call this `vertical_gap`?